File: 3.x.x/crates-utopia/queue/src/consumer.rs
use crate::error::QueueError;
use crate::message::Message;
use crate::publisher::Publisher;
use crate::queue::Queue;

/// Claim, ack, reject, and close a consumer.
///
/// PHP `Utopia\Queue\Consumer`.
pub trait Consumer: Send + Sync {
    /// Block up to `timeout` seconds for the next message and claim it, or `None` on timeout.
    fn receive(&self, queue: &Queue, timeout: i64) -> Result<Option<Message>, QueueError>;

    /// Acknowledge a processed message.
    fn commit(&self, queue: &Queue, message: &Message) -> Result<(), QueueError>;

    /// Mark a message as failed.
    fn reject(&self, queue: &Queue, message: &Message) -> Result<(), QueueError>;

    /// Close the consumer and free resources.
    fn close(&self);

    /// PHP `instanceof Publisher` check used by queue-depth telemetry.
    fn as_publisher(&self) -> Option<&dyn Publisher> {
        None
    }
}

impl Consumer for Box<dyn Consumer> {
    fn receive(&self, queue: &Queue, timeout: i64) -> Result<Option<Message>, QueueError> {
        (**self).receive(queue, timeout)
    }

    fn commit(&self, queue: &Queue, message: &Message) -> Result<(), QueueError> {
        (**self).commit(queue, message)
    }

    fn reject(&self, queue: &Queue, message: &Message) -> Result<(), QueueError> {
        (**self).reject(queue, message)
    }

    fn close(&self) {
        (**self).close();
    }

    fn as_publisher(&self) -> Option<&dyn Publisher> {
        (**self).as_publisher()
    }
}

impl Consumer for std::sync::Arc<dyn Consumer> {
    fn receive(&self, queue: &Queue, timeout: i64) -> Result<Option<Message>, QueueError> {
        (**self).receive(queue, timeout)
    }

    fn commit(&self, queue: &Queue, message: &Message) -> Result<(), QueueError> {
        (**self).commit(queue, message)
    }

    fn reject(&self, queue: &Queue, message: &Message) -> Result<(), QueueError> {
        (**self).reject(queue, message)
    }

    fn close(&self) {
        (**self).close();
    }

    fn as_publisher(&self) -> Option<&dyn Publisher> {
        (**self).as_publisher()
    }
}

File: src/Queue/Consumer.php
<?php

declare(strict_types=1);

namespace Utopia\Queue;

interface Consumer
{
    /**
     * Block up to $timeout seconds for the first message, then claim up to $n
     * messages without waiting for the batch to fill. Values below one use one.
     * Every returned message needs its own commit() or reject(). A failed claim
     * must leave removed messages recoverable, including after process death.
     *
     * @return list<Message> Empty on timeout.
     */
    public function receive(Queue $queue, int $timeout, int $n = 1): array;

    /** Acknowledge a processed message. */
    public function commit(Queue $queue, Message $message): void;

    /**
     * Mark a message as failed.
     *
     * The broker decides what a failure buys — another attempt, or the end of
     * the line — and this is where it decides it. A message the handler marked
     * terminal ({@see Message::terminal()}, set for it by a
     * {@see PermanentFailure}) must not be delivered again: park it wherever
     * exhausted messages go. A broker with no such notion can ignore the flag
     * and reject the way it always has.
     */
    public function reject(Queue $queue, Message $message): void;

    /** Close the consumer and free resources. */
    public function close(): void;
}
